### ISSUE TITLE
Queries which contain null/undefined values should not return the first/all documents

### ISSUE THREAD

**Issue by 8auu** [prisma/prisma#14976]
## Problem
Hello, 

Currently when trying to run a Prisma query if the value which you are using to find the documents for is `undefined` or `null`, then Prisma will return either the first document (If using a findFirst query) or all documents if you are using a query such as `.count`.

In the code snippet below, if `findGiveaway` is undefined, the query will return a count of the entire database table.
Code:
```
  const entriesCount = await client.prisma.giveawayEntry.count({
    where: {
      giveawayId: findGiveaway?.id,
    },
  });
```

Another example, which is a little more extreme can be seen below, in this example if `findUser` is undefined, it will find the first document in the table. In this use case this could mean that private user data could be return to the wrong user. 
```
   const findUserInfo = await client.prisma.userInfo.findFirst({
    where: {
      userId: findUser?.id,
    },
  });
```

## Suggested solution

Obviously there are ways around the issue explained above, by checking if the findUser is null/undefined before the query, but it would be great for an option inside of the query its self which could be `allowNull`, which by default could be true/false, maybe a possible option in `schema.prisma` to change the default value?

```
   const findUserInfo = await client.prisma.userInfo.findFirst({
    where: {
      userId: findUser?.id,
    },
    allowNull: false, // this would mean that if findUser is null/undefined it then the query will return a null value.
  });
```

**Comment by garrensmith** [comment]
Hey @8auu  this is by design. We have a section in our documentation https://www.prisma.io/docs/concepts/components/prisma-client/null-and-undefined that should explain this a but more.

**Comment by 8auu** [comment]
> Hey @8auu this is by design. We have a section in our documentation https://www.prisma.io/docs/concepts/components/prisma-client/null-and-undefined that should explain this a but more.

Hey, thanks for the link. Just edited the feature request to only mention null values.

**Comment by a-r-d** [comment]
I think this can be an incredible footgun and can cause insidious bugs and I will explain why.

When I query by an id on findFirst like this:

```
// somehow this id became undefined. 
const id = undefined

const user = await prisma.user.findFirst({
      where: {
        id: id,
      },
      rejectOnNotFound: true,
    })
```

Even when I do `rejectOnNotFound` if `id` accidentally at some point becomes set to undefined this comes back with a random object from the database that almost certainly not what I wanted to happen, in fact it probably causes a terrible bug in this case. 

I would expect `user` to come back null if I passed an `id` that does not match in a where clause but instead it omits the where clause entirely and I get something random.

The issue here is that types in typescript are erasable, they can be wrong at runtime and you can easily get something that is actually undefined even if it was typed as string.

**Comment by 8auu** [comment]
I completely agree with the above comment. Even now, knowing that this is how Prisma works I still almost just pushed a really really bad bug into production. Since finding this out last year I always try adding guard clauses where possible, but occasionally I forget.

In this instance, I am using NextAuth and Prisma. 

The types for session.user indicate that `id` should always be defined. For reasons currently unknown, this is not the case and it is returning `undefined` despite NextAuth session saying otherwise. 
```typescript
      console.log(session.user.id); // undefined
      if (session.user) {
        const u = await prisma.account.findFirst({
          where: {
            userId: session.user.id,
          },
        });
        ```

This shows that TypeScript types are not always 100% reliable, which almost certainly will lead to some pretty dangerous situations. If this were to be left unfound, anyone logging into my site would instantly gain access to the first accounts accessToken, which would be used for all future API requests.

**Comment by subvertallchris** [comment]
This keeps biting me. An undefined ID will slip past TypeScript and step right through an authorization check or query. The distinction between `undefined` and `null` makes sense but it would be great to have a way to tell Prisma to treat undefined as an error.

In the meantime, setting default arguments in repository functions is a quick and lazy way to ensure an omitted value will result in zero results:

```
export async function getArtistByIdAdmin({ id = '-1', userId = '-1' }: { id: string; userId: string });
```

TypeScript generously allows such code. Despite the fact that the type asserts that neither `id` nor `userId` will be undefined, we can set defaults at runtime in case we're wrong.
